$$
\newcommand{\F}{\mathbb{F}}
\newcommand{\R}{\mathbb{R}}
\newcommand{\a}{\mathbf{a}}
\newcommand{\b}{\mathbf{b}}
\newcommand{\betaa}{\mathbf{\beta}}
\newcommand{\c}{\mathbf{c}}
\newcommand{\u}{\mathbf{u}}
\newcommand{\v}{\mathbf{v}}
\newcommand{\w}{\mathbf{w}}
\newcommand{\x}{\mathbf{x}}
\newcommand{\y}{\mathbf{y}}
\newcommand{\z}{\mathbf{z}}
\newcommand{\A}{\mathbf{A}}
\newcommand{\B}{\mathbf{B}}
\newcommand{\C}{\mathbf{C}}
\newcommand{\X}{\mathbf{X}}
\newcommand{\rank}{\textbf{rank}}
\newcommand{\0}{\mathbf{0}}
\newcommand{\1}{\mathbf{1}}
\newcommand{\U}{\mathrm{U}}
\newcommand{\V}{\mathrm{V}}
\newcommand{\W}{\mathrm{W}}
\newcommand{\L}{\mathcal{L}}
\newcommand{\P}{\mathbb{P}}
$$

## Why is Softmax Layer Monotonic? 

- https://stackoverflow.com/questions/56096598/is-softmax-used-when-only-the-most-probable-class-will-be-used#:~:text=But%20softmax%20is%20a%20monotonic,layer%2C%20leaving%20the%20softmax%20out.

## Data Imbalance

- https://machinelearningmastery.com/random-oversampling-and-undersampling-for-imbalanced-classification/

### Robinhood Question [^Kevin Huo & Nick Singh: Ace The Data Science Interview, 2021. (pp. 118)]

- First, check if we can collect more data for the minority class, this may not always be feasible, but worth a try to ask first.
- Make sure the metrics defined for such problem is sensible, for example, using accuracy is not good and one should explain why this is the case by using the `ZeroR` example. One should also mention that AUROC, precision-recall metrics that are not so misleading.
- Over/Under sampling: delete majority class randomly till convergence to desired distribution, or resample with replacement minority class,
- Smote

[^Kevin Huo & Nick Singh: Ace The Data Science Interview, 2021. (pp. 118)]: **Kevin Huo & Nick Singh: Ace The Data Science Interview, 2021. (pp. 118)**

## Bias-Variance Tradeoff

- Definition
- Intuition and Examples
- My code on understanding the math and stats behind.

### Association with Overfit-Underfit

### How to tackle Bias-Variance Tradeoff

In the Ace the data science interview book, author mentioned how to approach case specific question.

Example

- high bias -> increase model complexity
- high var -> increase data etc

## Overfitting and Underfitting

### What is overfitting?

- https://sebastianraschka.com/faq/docs/overfitting.html

### How to tackle Overfitting?

#### Regularization

...


## Regularization

### Briefly desribe Regularization in Logistic Regression

- https://sebastianraschka.com/faq/docs/regularized-logistic-regression-performance.html

The visualization in the example is insane coupled with clear and intuitive explanation! Initially, you see the global minimum of the convex function is in the center of the graph, which is unregularized, we want to make the model a bit less "accurate" so we can have it generalize, so you see the small circle there is the "L2-term", now you can still find minimum, but with constraint to be in the small L2 circle.

### Why is the L1 constraint a diamond and L2 constraint a circle?

Consider a naive vector $\v = \begin{bmatrix} x \\ y \end{bmatrix} \in \R^2$. Assume all vectors start from the origin.

Define L2 norm as $\Vert \v \Vert_2 = \sqrt{x^2 + y^2}$ and then the **set of vectors (coordinates)** that satisfies this definition of L2 norm can be rewritten as:

$$
\left\{(x, y) ~|~ x^2+y^2 = r^2 \right\}
$$

where this definition means that if $\Vert \v \Vert_2 = 1$, then all vectors $\v$ having this norm must lie on the unit circle.


See the image from ISLR, consider a simple linear equation with two independent variables $\y = \beta_1 \x_1 + \beta_2 \x_2$, then the beta vector $\betaa = \begin{bmatrix} \beta_1 \\ \beta_2 \end{bmatrix}$ not has to fulfil a constraint such that say for example L2 norm (ridge) of $\betaa$ must be less than or equals to $r = 1$. Geomtrically, you can see the circle that the $\betaa$ must fall within and note that it may or may not contain the "true optimal $\betaa$".

<img src="https://storage.googleapis.com/reighns/reighns_ml_projects/docs/interview/misc_and_untidied/reg.PNG" style="margin-left:auto; margin-right:auto"/>
<p style="text-align: center">
    <b>ISLR</b>
</p>


**See pp.242-244 from ISLR for more intuition**.

- https://medium.com/uwaterloo-voice/a-deep-dive-into-regularization-eec8ab648bce
- https://www.mathworks.com/matlabcentral/answers/347583-how-to-plot-the-l2-norm-circle
- https://medium.com/@kiprono_65591/regularization-a-technique-used-to-prevent-over-fitting-886d5b361700
- https://www.analyticsvidhya.com/blog/2017/06/a-comprehensive-guide-for-linear-ridge-and-lasso-regression/

## Preprocessing Techniques

- should also check cs3244.

### Feature Scaling

- An article on which models may need scale: https://www.thekerneltrip.com/statistics/when-scale-my-data/
- Sebastian god: https://sebastianraschka.com/Articles/2014_about_feature_scaling.html
- More math: https://stats.stackexchange.com/questions/29781/when-conducting-multiple-regression-when-should-you-center-your-predictor-varia
- https://stats.stackexchange.com/questions/48360/is-standardization-needed-before-fitting-logistic-regression says If you use logistic regression with LASSO or ridge regression (as Weka Logistic class does) you should. As Hastie,Tibshirani and Friedman points out (page 82 of the pdf or at page 63 of the book): The ridge solutions are not equivariant under scaling of the inputs, and so one normally standardizes the inputs before solving.

### Encoding

- https://stats.stackexchange.com/questions/288095/what-algorithms-require-one-hot-encoding

## Model Selection, Evaluation

- Must read: Model Evaluation, Model Selection, and Algorithm Selection in Machine Learning by Seb

Some summary from the paper.

- 1.1 Performance Estimation: Generalization Performance vs. Model Selection
    - Why evaluate the predictive performance of a model?
    - Relative performance is accurate if we assume bias affects all the models we trained.
    - For example, if we trained 3 models, $M_1, M_2, M_3$, and their validation accuracy is ranked as $M_2: 75\% > M_1: 70\% > M_3: 65\%$, then if we add pessismistic bias of $10\%$, the three models still ranked in the same order relatively: $M_2: 65\% > M_1: 60\% > M_3: 55\%$. 
    - But do not think a $65\%$ validation accuracy translates to $65\%$ accuracy on the unseen test set.
    
- 1.2 Assumptions and Terminology
- 1.5 Holdout validation: v impt idea
- 3.9: Talks about whether feature selection in cv loop or out, data leak or ok.
- section 4: Hypothesis testing and algorithm comparison!

### What is Model Selection?

The process of finding the best-performing model from a set of models that were produced by different hyperparameter settings is called model selection.



## Gradient Descent

- https://realpython.com/gradient-descent-algorithm-python/

### Batch Gradient Descent

### Stochastic Gradient Descent

### Exploding and Vanishing Gradients

- https://www.analyticsvidhya.com/blog/2021/06/the-challenge-of-vanishing-exploding-gradients-in-deep-neural-networks/

## Metrics

### What models produce well calibrated probabilities?

- https://stats.stackexchange.com/questions/208867/why-does-logistic-regression-produce-well-calibrated-models

### Why do we sometimes use "multi-head" to evaluate multi-label classification?

- https://debuggercafe.com/multi-head-deep-learning-models-for-multi-label-classification/

### When should one use Sigmoid or Softmax in a Neural Network?

> [Sigmoid vs Softmax](https://stats.stackexchange.com/questions/233658/softmax-vs-sigmoid-function-in-logistic-classifier) I've noticed people often get directed to this question when searching whether to use sigmoid vs softmax in neural networks. If you are one of those people building a neural network classifier, here is how to decide whether to apply sigmoid or softmax to the raw output values from your network:

- If you have a multi-label classification problem = there is more than one "right answer" = the outputs are NOT mutually exclusive, then use a sigmoid function on each raw output independently. The sigmoid will allow you to have high probability for all of your classes, some of them, or none of them. Example: classifying diseases in a chest x-ray image. The image might contain pneumonia, emphysema, and/or cancer, or none of those findings.

---

- If you have a multi-class classification problem = there is only one "right answer" = the outputs are mutually exclusive, then use a softmax function. The softmax will enforce that the sum of the probabilities of your output classes are equal to one, so in order to increase the probability of a particular class, your model must correspondingly decrease the probability of at least one of the other classes. Example: classifying images from the MNIST data set of handwritten digits. A single picture of a digit has only one true identity - the picture cannot be a 7 and an 8 at the same time.

---

```python
# micro-averaged AUC
roc_auc_score(y_true.flatten(), y_pred.flatten())
# macro-averaged AUC
np.mean([roc_auc_score(y_true[:, i], y_pred[:, i]) for i in range(11)])
```

- https://scikit-learn.org/stable/modules/multiclass.html#multilabel-classification
- https://glassboxmedicine.com/2019/05/26/classification-sigmoid-vs-softmax/
- https://scikit-learn.org/stable/modules/multiclass.html#multilabel-classification

## Probability and Statistics

### Why do you multiply probabilities? 

Also appear in Chapter 2 summary.

Consider rolling a dice twice, what is the probability that you roll a 5 and 6 respectively.

We all know the answer is $\dfrac{1}{6} \times \dfrac{1}{6} = \dfrac{1}{36}$. But why?

This can be first understood that our denominator is the total outcomes in our sample space $\S$. This is $36$, why? By our counting principle on multiplication, we know that if we have $6$ choices in roll $1$ and $6$ choices in roll 2, then the cross-product is $6 \times 6 = 36$ total choices. One can enumerate $\{(1,1), (1,2), \ldots, (6,6)\}$ to see why.

Now the numerator is also related to the counting principle of multiplication as well! In roll 1, rolling a 5 is 1 choice, rolling a 6 next is 1 choice, so total there is a only one combination choice $1 \times 1$!

Now if we reframe the problem to what is the probability that you roll a 1, 2 or 3 in the first roll and 2 or 3 in the second roll. Then of course our denominator don't change as $36$, but our numerator changes, since in roll 1 we have 3 choices, and roll 2 have 2 choices, by the multiplicative principle we have a total of $3 \times 2 = 6$ choices, and so our probability is $\dfrac{6}{36}$ now. You can verify that there are indeed $6$ choices manually.

> **Now the most important part is we can use this if both events are independent! If not we need to be careful!**.

### What is Priori, Posterior and Conditional

Common terms in ML world.

Suppose there are three types of players in a tennis tournament: A, B, and C. Fifty percent of the contestants in the tournament are A players, 25% are B players, and 25% are C players. Your chance of beating the contestants depends on the class of the player, as follows:

- 0.3 against an A player
- 0.4 against a B player
- 0.5 against a C player

If you play a match in this tournament, what is the probability of your winning the
match? Supposing that you have won a match, what is the probability that you played
against an A player?

Let $W$ be the event that you win and $A$ be the event that you played vs player $A$, then

- Conditional: $\P(W~|~A)$ = given you played player $A$, what is your probability of winning?
- Priori: $\P(A)$ = **without entering the game**, what is your probability of facing player $A$?
- Posterior: $\P(A~|~W)$ = **after entering the game and winning the match**, what is your probability that you have actually played with $A$?
- Machine Learning: In many practical engineering problems, the question of interest is often the last one. That
is, supposing that you have observed something, what is the most likely cause of that event?
For example, supposing we have observed this particular dataset, what is the best Gaussian
model that would fit the dataset? Questions like these require some analysis of conditional
probability, prior probability, and posterior probability.


## Generic Questions

Usually generic and fundamental questions. These questions are "less code oriented", and usually requires a short and sweet answer, alongside with an example to illustrate.

### What are common pitfalls that lead to **Data Leakage** in a pipeline?

- https://scikit-learn.org/stable/common_pitfalls.html
- See also numerous question I asked on cross-validated.

### What is the meaning of "Train/Test" Distribution in the context of Machine Learning?

- https://towardsdatascience.com/understanding-entropy-the-golden-measurement-of-machine-learning-4ea97c663dc3#:~:text=By%20using%20entropy%20in%20machine,be%20desired%20in%20model%2Dbuilding.
- https://www.kdnuggets.com/2019/01/when-your-training-testing-data-different-distributions.html


Let us talk this in a less formal way without invoking formal definitions of "probability distributions".

#### Train and Test Distribution

Let us say we are imposed a Machine Learning problem to predict cats vs dogs. We are handed a training set, and there is also a hidden test set that we do not know of. Let us pretend we browse through each image in the training set $\mathcal{D}$, and plot out the pixel distribution of each image, we observe that each image pixel distribution follows a gaussian curve with mean 0 and std 1 (i.e. each image in $\mathcal{D}$ is generated from $\mathcal{N}(0,1)$). Then we trained a good model using training set, but performed badly on the test set $\mathcal{D}_{test}$, and after we also plot the pixel density of the test set, we discovered that this set follows a very different distributions, say follow $\mathcal{N}(255, 255)$, and each image has 10 times the resolution of the training set. We can therefore conclude, that, both train and test set have different "probability distributions".

---

#### Distribution from your model

Let us also continue our example from the cat vs dog, this time we assume both train, test and every other images for dogs and cats are generated from $\mathcal{N}(0, 1)$. That is to say, we **know** our population distribution, and if we do know, then why do we need a Machine Learning model at all? We can just approximate all images with the known population distribution with the parameters. Even simpler, if we have a dataset that is beknownst to us with a perfect linear relationship that can be predicted perfectly by a simple linear regression model, (i.e. the population parameter is known), then why do we even need linear regression?

However, this world is far from perfect, and we do not know the population parameters of the underlying distribution of the dataset given to us, hence we use models to predict it.

Thus, we can analogously say that if a training set $\mathcal{D}$ follows a distribution $p$, then our model follows a distribution $q$, where in the simplest case in linear regression, we know our $q$ follows a distribution with 2 parameters, $m$ the slope, $c$ the bias.

With that in mind, our machine learning models can be phrased as using $q$ to approximate $q$. And this is often used in Cross-Entropy loss.

### Parametric and Non-Parametric Models.

- ISLR (p21-24)
- https://stats.stackexchange.com/questions/268638/what-exactly-is-the-difference-between-a-parametric-and-non-parametric-model
- https://sebastianraschka.com/faq/docs/parametric_vs_nonparametric.html

> https://sebastianraschka.com/faq/docs/parametric_vs_nonparametric.html:
> The term “non-parametric” might sound a bit confusing at first: non-parametric does not mean that they have NO parameters! On the contrary, non-parametric models (can) become more and more complex with an increasing amount of data.

> So, in a parametric model, we have a finite number of parameters, and in nonparametric models, the number of parameters is (potentially) infinite. Or in other words, in nonparametric models, the complexity of the model grows with the number of training data; in parametric models, we have a fixed number of parameters (or a fixed structure if you will).

> Linear models such as linear regression, logistic regression, and linear Support Vector Machines are typical examples of a parametric “learners;” here, we have a fixed size of parameters (the weight coefficient.) In contrast, K-nearest neighbor, decision trees, or RBF kernel SVMs are considered as non-parametric learning algorithms since the number of parameters grows with the size of the training set. – K-nearest neighbor and decision trees, that makes sense, but why is an RBF kernel SVM non-parametric whereas a linear SVM is parametric? In the RBF kernel SVM, we construct the kernel matrix by computing the pair-wise distances between the training points, which makes it non-parametric.

> In the field of statistics, the term parametric is also associated with a specified probability distribution that you “assume” your data follows, and this distribution comes with the finite number of parameters (for example, the mean and standard deviation of a normal distribution); you don’t make/have these assumptions in non-parametric models. So, in intuitive terms, we can think of a non-parametric model as a “distribution” or (quasi) assumption-free model.

> However, keep in mind that the definitions of “parametric” and “non-parametric” are “a bit ambiguous” at best; according to the “The Handbook of Nonparametric Statistics 1 (1962) on p. 2: “A precise and universally acceptable definition of the term ‘nonparametric’ is not presently available. The viewpoint adopted in this handbook is that a statistical procedure is of a nonparametric type if it has properties which are satisfied to a reasonable approximation when some assumptions that are at least of a moderately general nature hold."

---

### Supervised and Unsupervised Learning.

- ISLR (p26-28)

### Tradeoff between Flexibility and Interpretability of Machine Learning Models.

- ISLR (p24-26)

### Reducible and Irreducible Errors.

- ISLR (p18-19)

#### Definition

Formal Definition

#### Example (Intuition)

- We can use features $\X$ to predict $\y$, for example, use **age of house, number of bedrooms, etc** to predict the **price of the house**.
- The **irreducible errors** can be random events such as Elon Musk tweeting something good about a particular house, which our model cannot account for.

### Describe Ensembling in laymen terms.

**François Chollet**(Creator of Keras) explained it as : Ensembling relies on the assumption that different good models trained independently are likely to be good for different reasons: each model looks at slightly different aspects of the data to make its predictions, getting part of the “truth” but not all of it. You may be familiar with the ancient parable of the blind men and the elephant: a group of blind men come across an elephant for the first time and try to understand what the elephant is by touching it. Each man touches a different part of the elephant’s body—just one part, such as the trunk or a leg. Then the men describe toeach other what an elephant is: “It’s like a snake,” “Like a pillar or a tree,” and so on.The blind men are essentially machine-learning models trying to understand the manifold of the training data, each from its own perspective, using its own assumptions(provided by the unique architecture of the model and the unique random weight initialization). Each of them gets part of the truth of the data, but not the whole truth. By pooling their perspectives together, you can get a far more accurate description of the data. The elephant is a combination of parts: not any single blind man gets it quiteright, but, interviewed together, they can tell a fairly accurate story.

Let’s use classification as an example. The easiest way to pool the predictions of a setof classifiers (to ensemble the classifiers) is to average their predictions at inference time:Use four different models to compute initial predictions.

```
preds_a = model_a.predict(x_val)
preds_b = model_b.predict(x_val)
preds_c = model_c.predict(x_val)
preds_d = model_d.predict(x_val)

```

This new prediction array should be more accurate than any of the initial ones.

`final_preds = 0.25 * (preds_a + preds_b + preds_c + preds_d)`

This will work only if the classifiers are more or less equally good. If one of them is significantly worse than the others, the final predictions may not be as good as the best classifier of the group.A smarter way to ensemble classifiers is to do a weighted average, where the weights are learned on the validation data typically, the better classifiers are given a higher weight, and the worse classifiers are given a lower weight. To search for a good set of ensembling weights, you can use random search or a simple optimization algorithm such as Nelder-Mead:

```
preds_a = model_a.predict(x_val)
preds_b = model_b.predict(x_val)
preds_c = model_c.predict(x_val)
preds_d = model_d.predict(x_val)

```

These weights (0.5, 0.25,0.1, 0.15) are assumed to be learned empirically.`final_preds = 0.5 * preds_a + 0.25 * preds_b + 0.1 * preds_c + 0.15 * preds_d`

## Ensembling Theory

### Bagging

## Cross-Validation

### Why does Cross-Validation reduce Variance?

## Model Evaluation

- https://www.ritchieng.com/applying-machine-learning/

## Linear Regression

### Polynomial Regression

- https://zerowithdot.com/polynomial-regression-in-python/

## Logistic Regression

## Decision Tree

### Describe DT briefly

- Suprvised Learning: will need corresponding label
- Non-parametric
- 

## Convolutional Neural Networks

In [6]:
# !pip install torchinfo
import timm
import torch
import torchvision
from typing import Dict, Union, Callable, OrderedDict, Tuple
import os, random
import numpy as np
import torch.nn as nn
from torchinfo import summary

def seed_all(seed: int = 1992) -> None:
    """Seed all random number generators."""
    print(f"Using Seed Number {seed}")

    os.environ["PYTHONHASHSEED"] = str(seed)  # set PYTHONHASHSEED env var at fixed value
    torch.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    torch.cuda.manual_seed(seed)  # pytorch (both CPU and CUDA)
    np.random.seed(seed)  # for numpy pseudo-random generator
    # set fixed value for python built-in pseudo-random generator
    random.seed(seed)
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = True
    torch.backends.cudnn.enabled = True


def seed_worker(_worker_id) -> None:
    """Seed a worker with the given ID."""
    worker_seed = torch.initial_seed() % 2 ** 32
    np.random.seed(worker_seed)
    random.seed(worker_seed)


seed_all(seed=1992)

Using Seed Number 1992


### Padding

#### What is the role of zero padding?



### Pooling

#### Global Average Pooling

- https://rwightman.github.io/pytorch-image-models/feature_extraction/#penultimate-layer-features-pre-classifier-features

In [65]:
import torch
import timm

model = timm.create_model('resnet18', pretrained=True, num_classes=1000)
batch_size = 2
image_shape = (3, 224, 224)
input_image_tensor = torch.rand(size=(batch_size, *image_shape))


# print(summary(model, (2, 3, 224, 224)))

- `model.reset_classifier(num_classes=0, global_pool="")` means we **do not want global average pooling** and thus the shape at the last conv layer (penultimate layer) is $(2, 512, 7, 7)$
- `model.reset_classifier(num_classes=0, global_pool="avg")` means we **do want global average pooling** and thus the shape at the last conv layer (penultimate layer) is $(2, 512)$
   whereby for each and every of the 512 feature maps $f_i$, we average $f_i$ across all pixels (i.e. if $f_i$ is 3 by 3 then average means add all $3 \times 3 = 9$ pixels and average) and concat to become one $512$ vector. 

In [66]:
model = timm.create_model('resnet18', pretrained=True, num_classes=1000)

o = model(input_image_tensor)
print(f'Original shape: {o.shape}')

model.reset_classifier(num_classes=0, global_pool="")

o = model(input_image_tensor)
print(f'Unpooled shape: {o.shape}')

Original shape: torch.Size([2, 1000])
Unpooled shape: torch.Size([2, 512, 7, 7])


In [67]:
model = timm.create_model('resnet18', pretrained=True, num_classes=1000)

o = model(input_image_tensor)
print(f'Original shape: {o.shape}')

model.reset_classifier(num_classes=0, global_pool="avg")

o = model(input_image_tensor)
print(f'Pooled shape: {o.shape}')

Original shape: torch.Size([2, 1000])
Pooled shape: torch.Size([2, 512])


### Image Normalization in Transfer Learning

[Image Normalization](https://reighns92.github.io/reighns-ml-blog/reighns_ml_journey/machine_learning_and_deep_learning/computer_vision/image_normalization/Image_Normalization_and_Standardization/)

### 1x1 Convolution

- [A Gentle Introduction to 1×1 Convolutions to Manage Model Complexity](https://machinelearningmastery.com/introduction-to-1x1-convolutions-to-reduce-the-complexity-of-convolutional-neural-networks/)

#### Intuition

Pooling layers are designed to downscale **feature maps** and systematically halve the **width and height** of feature maps in the network. Nevertheless, pooling layers do not change the number of filters in the model which is the **depth or number of channels**. The intuition of 1x1 convolution is that we can think of it as a **downsample** operation on the **feature maps** but this time shrinking the **depth/channels** instead of the **width/height**.

#### Downsample Feature Maps with 1x1 Convolution

In Andrew Ng's example, let us define:

- Batch size of 1;
- A conv layer of with kernel size of 5, same padding of 2 and stride of 1, and an output depth of 32;
- The input shape is a stack of feature maps of 192 depth/channels, each of 28 by 28;
- It follows that the learnable params of the this conv layer is $32 \times 5 \times 5 \times 192 + 32 = 153, 632$.  
- It follows that the output shape is of $(32, 28, 28)$, which is a stack of feature maps of 32 depth, each of 28 by 28 size.

In [18]:
model = nn.Sequential(
    nn.Conv2d(
        in_channels=192, out_channels=32, kernel_size=5, padding=2, stride=1
    ),

)
batch_size = 1
input_image = torch.rand(size=(192, 28, 28))

print(summary(model, (batch_size, *input_image.shape)))

Layer (type:depth-idx)                   Output Shape              Param #
Sequential                               --                        --
├─Conv2d: 1-1                            [1, 32, 28, 28]           153,632
Total params: 153,632
Trainable params: 153,632
Non-trainable params: 0
Total mult-adds (M): 120.45
Input size (MB): 0.60
Forward/backward pass size (MB): 0.20
Params size (MB): 0.61
Estimated Total Size (MB): 1.42


Now if we design a **intermediate** layer that holds 1x1 convolution that first decreases the depth from 192 to 16 instead of 32, then connect back with another conv layer of kernel size 5 and depth 32, we can recover the same output shape but at a much leseer computational cost. As we can see below, the params reduced almost 10 folds.

Commonly, we call this **intermediate** layer a **bottleneck layer**.

In [20]:
model = nn.Sequential(
    nn.Conv2d(
        in_channels=192, out_channels=16, kernel_size=1, padding=0, stride=1
    ),
    nn.Conv2d(
        in_channels=16, out_channels=32, kernel_size=5, padding=2, stride=1
    ),
)
batch_size = 1
input_image = torch.rand(size=(192, 28, 28))

print(summary(model, (batch_size, *input_image.shape)))

Layer (type:depth-idx)                   Output Shape              Param #
Sequential                               --                        --
├─Conv2d: 1-1                            [1, 16, 28, 28]           3,088
├─Conv2d: 1-2                            [1, 32, 28, 28]           12,832
Total params: 15,920
Trainable params: 15,920
Non-trainable params: 0
Total mult-adds (M): 12.48
Input size (MB): 0.60
Forward/backward pass size (MB): 0.30
Params size (MB): 0.06
Estimated Total Size (MB): 0.97


#### Upsample Feature Maps with 1x1 Convolution

Similarly, we can increase the **depth/channel** of the **feature maps** while maintaining the **width and height**.

#### 1x1 Convolution is Equivalent to a FC-layer

### Downsampling

#### Convolutional Stride (Downsampling)

In [33]:
batch_size = 1
input_image = torch.rand(size=(3, 224, 224))

In [36]:
model = nn.Sequential(
    nn.Conv2d(
        in_channels=3, out_channels=64, kernel_size=7, padding=3, stride=1, bias=False
    ),
)


print(summary(model, (batch_size, *input_image.shape)))

Layer (type:depth-idx)                   Output Shape              Param #
Sequential                               --                        --
├─Conv2d: 1-1                            [1, 64, 224, 224]         9,408
Total params: 9,408
Trainable params: 9,408
Non-trainable params: 0
Total mult-adds (M): 472.06
Input size (MB): 0.60
Forward/backward pass size (MB): 25.69
Params size (MB): 0.04
Estimated Total Size (MB): 26.33


In [37]:
model = nn.Sequential(
    nn.Conv2d(
        in_channels=3, out_channels=64, kernel_size=7, padding=3, stride=2, bias=False
    ),
)


print(summary(model, (batch_size, *input_image.shape)))

Layer (type:depth-idx)                   Output Shape              Param #
Sequential                               --                        --
├─Conv2d: 1-1                            [1, 64, 112, 112]         9,408
Total params: 9,408
Trainable params: 9,408
Non-trainable params: 0
Total mult-adds (M): 118.01
Input size (MB): 0.60
Forward/backward pass size (MB): 6.42
Params size (MB): 0.04
Estimated Total Size (MB): 7.06


This is considered a **downsampling** operation due to our stride changed from 2 to 1. The output shape of the feature maps is from (64, 224, 224) to (64, 112, 112), effectively halving the width and height of the feature maps.

#### Pooling (Downsampling)

- https://www.quora.com/What-is-a-downsampling-layer-in-Convolutional-Neural-Network-CNN

<div class="container" style="display: inline-block;">  
  <figure>
  <div style="float: left; padding: 10px;">
    <img src='https://storage.googleapis.com/reighns/reighns_ml_projects/docs/deep_learning/computer_vision/data/misc/Illustration-of-Max-Pooling-and-Average-Pooling-Figure-2-above-shows-an-example-of-max.png' width="350" height="350" align="center"/>
    <figcaption align="center"><b>Pooling as a form of Downsampling. <a href="https://www.researchgate.net/figure/Illustration-of-Max-Pooling-and-Average-Pooling-Figure-2-above-shows-an-example-of-max_fig2_333593451">Source</a></b></figcaption>
  </div>
  </figure>
</div>


In [38]:
model = nn.Sequential(
    nn.Conv2d(
        in_channels=3, out_channels=64, kernel_size=7, padding=3, stride=1, bias=False
    )
)
print(summary(model, (batch_size, *input_image.shape)))

Layer (type:depth-idx)                   Output Shape              Param #
Sequential                               --                        --
├─Conv2d: 1-1                            [1, 64, 224, 224]         9,408
Total params: 9,408
Trainable params: 9,408
Non-trainable params: 0
Total mult-adds (M): 472.06
Input size (MB): 0.60
Forward/backward pass size (MB): 25.69
Params size (MB): 0.04
Estimated Total Size (MB): 26.33


In [39]:
model = nn.Sequential(
    nn.Conv2d(
        in_channels=3, out_channels=64, kernel_size=7, padding=3, stride=1, bias=False
    ),
    nn.MaxPool2d(kernel_size=2, stride=2),
)

print(summary(model, (batch_size, *input_image.shape)))

Layer (type:depth-idx)                   Output Shape              Param #
Sequential                               --                        --
├─Conv2d: 1-1                            [1, 64, 224, 224]         9,408
├─MaxPool2d: 1-2                         [1, 64, 112, 112]         --
Total params: 9,408
Trainable params: 9,408
Non-trainable params: 0
Total mult-adds (M): 472.06
Input size (MB): 0.60
Forward/backward pass size (MB): 25.69
Params size (MB): 0.04
Estimated Total Size (MB): 26.33


Spatial pooling layers such as `MaxPool2d` with a size of 2 and stride 2 also reduces the feature map by a factor of 2 by 2. 

Notice that both downsampling methods have the same number of **trainable parameters** and also the exact **same output shape**, so what's the difference?

> Maybe the additional `MaxPool2d` is an additional operation that is not needed? Let's see two examples.

#### Convolutional Stride vs Pooling for Downsampling

The advantage of the convolution layer is that it can learn certain properties that you might not think of while you add pooling layer. Pooling is a fixed operation and convolution can be learned. On the other hand, pooling is a cheaper operation than convolution, both in terms of the amount of computation that you need to do and number of parameters that you need to store (no parameters for pooling layer).

There are examples when one of them is better choice than the other.

##### Example when the convolution with strides is better than pooling

The first layer in the ResNet uses convolution with strides. This is a great example of when striding gives you an advantage. This layer by itself significantly reduces the amount of computation that has to be done by the network in the subsequent layers. It compresses multiple 3x3 convolution (3 to be exact) in to one 7x7 convolution, to make sure that it has exactly the same receptive field as 3 convolution layers (even though it is less powerful in terms of what it can learn). At the same time this layer applies stride=2 that downsamples the image. Because this first layer in ResNet does convolution and downsampling at the same time, the operation becomes significantly cheaper computationally. If you use stride=1 and pooling for downsampling, then you will end up with convolution that does 4 times more computation + extra computation for the next pooling layer. The same trick was used in SqueezeNet and some other neural network architectures.

##### Example where pooling is better than convolution

In the NIPS 2018, there was a new architecture presented called [FishNet][1]. One thing that they try is to fix the problems with the residual connections used in the ResNet. In the ResNet, in few places, they put 1x1 convolution in the skip connection when downsampling was applied to the image. This convolution layer makes gradient propagation harder. One of the major changes in their paper is that they get rid of the convolutions in the residual connections and replaced them with pooling and simple upscales/identities/concatenations. This solution fixes problem with gradient propagation in very deep networks. 

From the FishNet paper (Section 3.2)

> The layers in the head are composed of concatenation, convolution with
> identity mapping, and max-pooling. Therefore, the gradient propagation
> problem from the previous backbone network in the tail are solved with
> the FishNet by 1) excluding I-conv at the head; and 2) using
> concatenation at the body and the head.


  [1]: https://papers.nips.cc/paper/7356-fishnet-a-versatile-backbone-for-image-region-and-pixel-level-prediction.pdf
 
- [Pooling vs. stride for downsampling](https://stats.stackexchange.com/questions/387482/pooling-vs-stride-for-downsampling)

### Feature Maps

#### Why are the last few Feature Maps more Useful in Feature extraction?

- The backbone network ("convolution and pooling") is responsible for extracting a feature map from the image that contains higher level summarized information. Each head uses this feature map as input to predict its desired outcome. 
- The main intuition why feature maps of the last few layers (last layer usually) are important is one needs to recognize the earlier conv layer's feature maps find simple features like shapes, sizes, edges from an image, while the deep conv layers will be of more abstract features in an image. As a result, we really just want the abstract feature maps as they are **more class specific to the image** instead of the earlier layers which gives **generic shapes**.

## Object Detection (YOLO)

## Object Detection (Bounding Boxes)

- Different conversions
    - https://albumentations.ai/docs/getting_started/bounding_boxes_augmentation/
    - https://github.com/awsaf49/bbox/blob/main/bbox/utils.py